src/polkassembly/polkassembly.ts

```diff
@@ -0,0 +1,143 @@
+    private signer: { type: "polkadot"; keyringPair: KeyringPair } | { type: "ethereum"; wallet: Wallet }, // Ethereum type is used for EVM chains.
+  ) {}
+
+  public get loggedIn(): boolean {
+    return this.token !== undefined;
+  }
+
+  public get address(): string {
+    return this.signer.type === "polkadot" ? this.signer.keyringPair.address : this.signer.wallet.address;
+  }
+
+  public async signup(): Promise<void> {
+    if (this.loggedIn) return;
+    const signupStartResponse = await fetch(`${this.endpoint}/auth/actions/addressSignupStart`, {
+      headers,
+      method: "POST",
+      body: JSON.stringify({ address: this.address }),
+    });
+    if (!signupStartResponse.ok) {
+      throw new Error(await signupStartResponse.text());
+    }
+    const signupStartBody = (await signupStartResponse.json()) as { signMessage: string };
+
+    const signupResponse = await fetch(`${this.endpoint}/auth/actions/addressSignupConfirm`, {
+      headers,
+      method: "POST",
+      body: JSON.stringify({
+        address: this.address,
+        signature: await this.signMessage(signupStartBody.signMessage),
+        wallet: this.signer.type === "polkadot" ? "polkadot-js" : "metamask",
+      }),
+    });
+    if (!signupResponse.ok) {
+      throw new Error(await signupResponse.text());
+    }
+    const signupBody = (await signupResponse.json()) as { token: string };
+    if (!signupBody.token) {
+      throw new Error("Signup unsuccessful, the authentication token is missing.");
+    }
+    this.token = signupBody.token;
+  }
+
+  public async login(): Promise<void> {
+    if (this.loggedIn) return;
+
+    const loginStartResponse = await fetch(`${this.endpoint}/auth/actions/addressLoginStart`, {
+      headers,
+      method: "POST",
+      body: JSON.stringify({ address: this.address }),
+    });
+    if (!loginStartResponse.ok) {
+      throw new Error(await loginStartResponse.text());
+    }
+    const loginStartBody = (await loginStartResponse.json()) as { signMessage: string };
+
+    const loginResponse = await fetch(`${this.endpoint}/auth/actions/addressLogin`, {
+      headers,
+      method: "POST",
+      body: JSON.stringify({
+        address: this.address,
+        signature: await this.signMessage(loginStartBody.signMessage),
+        wallet: this.signer.type === "polkadot" ? "polkadot-js" : "metamask",
+      }),
+    });
+    if (!loginResponse.ok) {
+      throw new Error(await loginResponse.text());
+    }
+    const loginBody = (await loginResponse.json()) as { token: string };
+    if (!loginBody.token) {
+      throw new Error("Login unsuccessful, the authentication token is missing.");
+    }
+    this.token = loginBody.token;
+  }
+
+  public logout(): void {
+    this.token = undefined;
+  }
+
+  public async loginOrSignup(): Promise<void> {
+    try {
+      await this.login();
+    } catch (e) {
+      if ((e as Error).message.includes("Please sign up")) {
+      }
+      await this.signup();
```

Comment:
Shouldn't this be inside of the `if` statement?

Reply:
Right, of course - mistake corrected.